pageserver/src/tenant/mgr.rs

```diff
@@ -1517,21 +1521,148 @@ impl TenantManager {
+            let parent_path = parent_path.clone();
+
+            // Since we will do a large number of small filesystem metadata operations, batch them into
+            // spawn_blocking calls rather than doing each one as a tokio::fs round-trip.
+            let span = tracing::info_span!("link_layers", tenant_id=%child.tenant_id, shard_id=%child.shard_slug());
+            let jh = tokio::task::spawn_blocking(move || -> anyhow::Result<()> {
+                let _e = span.entered();
+                for dir in create_dirs {
+                    if let Err(e) = std::fs::create_dir_all(&dir) {
+                        // Ignore AlreadyExists errors, drop out on all other errors
+                        match e.kind() {
+                            std::io::ErrorKind::AlreadyExists => {}
+                            _ => {
+                                return Err(anyhow::anyhow!(e).context(format!("Creating {dir}")));
+                            }
+                        }
+                    }
+                }
+
+                for relative_layer in layers_clone {
+                    let parent_path = parent_path.join(&relative_layer);
+                    let child_path = child_prefix.join(&relative_layer);
+                    if let Err(e) = std::fs::hard_link(&parent_path, &child_path) {
+                        match e.kind() {
+                            std::io::ErrorKind::AlreadyExists => {}
+                            std::io::ErrorKind::NotFound => {
+                                tracing::info!(
+                                    "Layer {} not found during hard-linking, evicted during split?",
+                                    relative_layer
+                                );
+                            }
+                            _ => {
+                                return Err(anyhow::anyhow!(e).context(format!(
+                                    "Hard linking {relative_layer} into {child}"
+                                )))
+                            }
+                        }
+                    }
+                }
+
+                Ok(())
+            });
+
+            match jh.await {
+                Ok(Ok(())) => {
+                    tracing::info!(
+                        count = parent_layers.len(),
+                        tenant_id = %child.tenant_id,
+                        shard_id = %child.shard_slug(),
+                        "Hard-linked layers into child shard",
+                    );
+                }
+                Ok(Err(e)) => {
+                    // This is an optimization, so we tolerate failure.
+                    tracing::warn!("Error hard-linking layers, proceeding anyway: {e}")
+                }
+                Err(e) => {
+                    // This is something totally unexpected like a panic, so bail out.
+                    anyhow::bail!("Error joining hard linking task: {e}");
+                }
+            }
```

Comment:
changing this to use a joinset would be a matter of returning the shard_id from spawn_blocking task. But I cannot see why would hard linking be slow.

Why not do everything in single spawn_blocking? it'd be a matter of moving the for loop into it and save N hops.

Reply:
>Why not do everything in single spawn_blocking? it'd be a matter of moving the for loop into it and save N hops.

Makes sense, refactored in fcc2eb88a1a849793fbdde231f4c421013db735b